File: src/skybox.c
#include "skybox.h"

#include "raylib.h"
#include "rlgl.h"

Skybox skybox_generate(Texture tex) {
    Skybox s;

    s.cube = GenMeshCube(1.0, 1.0, 1.0);
    s.model = LoadModelFromMesh(s.cube);

    bool use_hdr = false;

    // NOTE: web GLSL version is 100 not 330
    s.model.materials[0].shader = LoadShader(
        "shaders/skybox.vs",
        "shaders/skybox.fs"
    );

    SetShaderValue(s.model.materials[0].shader, GetShaderLocation(s.model.materials[0].shader, "environmentMap"), (int[1]){ MATERIAL_MAP_CUBEMAP }, SHADER_UNIFORM_INT);
    SetShaderValue(s.model.materials[0].shader, GetShaderLocation(s.model.materials[0].shader, "doGamma"), (int[1]){ use_hdr? 1 : 0 }, SHADER_UNIFORM_INT);
    SetShaderValue(s.model.materials[0].shader, GetShaderLocation(s.model.materials[0].shader, "vflipped"), (int[1]){ use_hdr? 1 : 0 }, SHADER_UNIFORM_INT);
    
    // cubemap
    
    Shader cubemap_shader = LoadShader(
        "shaders/cubemap.vs",
        "shaders/cubemap.fs"
    );

    SetShaderValue(cubemap_shader, GetShaderLocation(cubemap_shader, "equirectangularMap"), (int[1]){ 0 }, SHADER_UNIFORM_INT);

    Image img = LoadImageFromTexture(tex);

    // apply image
    s.model.materials[0].maps[MATERIAL_MAP_CUBEMAP].texture = LoadTextureCubemap(img, CUBEMAP_LAYOUT_AUTO_DETECT);
    UnloadImage(img);

    return s;
}

void skybox_unload(Skybox* sb) {
    UnloadModel(sb->model);
    // UnloadMesh(sb->cube);
}

void skybox_draw(Skybox* sb) {
    rlDisableBackfaceCulling();
    rlDisableDepthMask();

    DrawModel(sb->model, (Vector3){0, 0, 0}, 1.0f, WHITE);

    rlEnableBackfaceCulling();
    rlEnableDepthMask();
}

File: src/skybox.rs
use raylib::{
    ffi::{
        rlDisableBackfaceCulling, rlDisableDepthMask, rlEnableBackfaceCulling, rlEnableDepthMask,
    },
    prelude::*,
};

pub struct Skybox {
    model: Model,
}

impl Skybox {
    pub fn new(rl: &mut RaylibHandle, thr: &RaylibThread, image: &Image) -> Self {
        let cube = unsafe { Mesh::gen_mesh_cube(&thr, 1.0, 1.0, 1.0).make_weak() };
        let mut model = rl.load_model_from_mesh(&thr, cube).unwrap();

        // --- skybox shader
        let mut skybox_shader =
            rl.load_shader(&thr, Some("shader/skybox.vs"), Some("shader/skybox.fs"));

        skybox_shader.set_shader_value(
            skybox_shader.get_shader_location("environmentMap"),
            &[MaterialMapIndex::MATERIAL_MAP_CUBEMAP as i32][..],
        );
        skybox_shader.set_shader_value(skybox_shader.get_shader_location("doGamma"), &[0][..]);
        skybox_shader.set_shader_value(skybox_shader.get_shader_location("vflipped"), &[0][..]);

        let skybox_shader = skybox_shader.to_raw();

        model.materials_mut()[0].shader = skybox_shader;

        // --- cubemap texture
        let cubemap_texture = rl
            .load_texture_cubemap(&thr, image, CubemapLayout::CUBEMAP_LAYOUT_AUTO_DETECT)
            .unwrap()
            .to_raw();

        dbg!(&cubemap_texture);

        model.materials_mut()[0].maps_mut()[MaterialMapIndex::MATERIAL_MAP_CUBEMAP as usize]
            .texture = cubemap_texture;

        Self { model }
    }

    pub fn draw(&self, rl: &mut impl RaylibDraw3D) {
        unsafe {
            rlDisableBackfaceCulling();
            rlDisableDepthMask();
        }
        rl.draw_model(&self.model, Vector3::zero(), 1.0, Color::WHITE);
        unsafe {
            rlEnableBackfaceCulling();
            rlEnableDepthMask();
        }
    }
}

impl Drop for Skybox {
    fn drop(&mut self) {
        unsafe {
            raylib::ffi::UnloadShader(self.model.materials_mut()[0].shader);
            raylib::ffi::UnloadTexture(
                self.model.materials_mut()[0].maps_mut()
                    [MaterialMapIndex::MATERIAL_MAP_CUBEMAP as usize]
                    .texture,
            );
        }
    }
}
